stm32_base/hal/timer_stm32/timer_device_stm32.c

```diff
@@ -377,26 +380,30 @@ void TIM1_BRK_TIM9_IRQHandler(void)
 
 void TIM2_IRQHandler(void)
 {
-    static uint32_t last = 0;
+    static uint32_t last;
+    static uint32_t sigs;
 
-    /* TODO add later when this can work */
-#if 0
     if (TIM_GetITStatus(TIM2, TIM_IT_Update) != RESET) {
         /* Overflow interrupt */
-        TIM_ClearITPendingBit(TIM2, TIM_IT_Update);
-        timer2_period = 0;
+            TIM_ClearITPendingBit(TIM2, TIM_IT_Update);
+
+            if (0 == sigs)
+                    timer2_period = 0;
+            sigs = 0;
     }
-#endif
+
     if (TIM_GetITStatus(TIM2, TIM_IT_CC4) != RESET) {
         /* THIS IS A HORRIBLE HACK. */
         /* TIM2 CH4 does not seem to reset the counter upon input capture. this is doing it manually. */
         /* need to fix the configuration of this timer, if possible. */
         uint32_t current = TIM_GetCapture4(TIM2);
         if (last < current) {
-            uint32_t delta = current - last;
-            timer2_period = delta;
+                timer2_period = current - last;
+        } else {
+                timer2_period = TIMER_PERIOD - last + current;
         }
         last = current;
+        ++sigs;
```

Comment:
Stands for signals.  It counts the number of times an interrupt occurs on a CC register